pycalphad/tests/test_equilibrium.py

```diff
@@ -703,6 +704,207 @@ def test_eq_associate():
     eq = equilibrium(dbf, ['A', 'Q'], phases, conds)
     assert_allclose(eq.GM.values, -1736.981311)
 
+@select_database("Ocadiz-Flores.dat")
+def test_MQMQA_energy_K_F_NI(load_database):
+    dbf = load_database()
+    comps = ['K', 'F', 'NI']
+    model = {'LIQUID2': ModelMQMQA}
```

Comment:
Anything we can do to avoid requiring users to explicitly pass `ModelMQMQA`?

Reply:
When I was working on the `ModelProtocol` implementation a while ago I came up with a `ModelDispatcher` class (a function would be fine too) that has the same API as `Model.__init__` that dispatches on the correct class depending on `model_hints`. A possible solution would be to have this class be the default argument for `unpack_kwarg` for the `"model"` kwarg. 

For context, I ultimately decided against this kind of solution because I thought it would be easier for users to be comfortable with passing non-`Model` models explicitly. The two options I came up with for being able to hook into this dispatcher weren't very good: 
1. Users register their custom models with pycalphad to be able to be picked up by the dispatcher, which is kind of an annoying gotcha and requires keeping state
2. The `ModelProtocol` that users would subclass uses metaclass magic to automatically register the class, so any subclass would be transparently registered for users, but this also created some state and seemed like too much magic and clunkiness.

Some other options (that, IMO, are worse than the manual or dispatcher approaches):
- Add logic to `calculate` or `equilibrium` to pick the right class based on `model_hints`. I'd be worried about implementing this in a way that doesn't override user-specified models - a solution that worked around this issue would probably just look like the dispatcher approach.
- Add some logic in the `Model` initializer to instantiate the `ModelMQMQA` class instead. I think it would be bad form to give a caller of `Model` a different class than the one they asked for.

Reply:
What about something like

```python
class Model:
    def __new__(cls, *args, **kwargs):
        target_cls = cls.dispatch_on(*args, **kwargs)
        instance = object.__new__(target_cls)
        return instance

    @classmethod
    def dispatch_on(cls, dbe, comps, phase_name, parameters=None):
        phase = dbe.phases[phase_name.upper()]
        target_cls = cls
        if 'mqmqa' in phase.model_hints.keys():
            from pycalphad.models.model_mqmqa import ModelMQMQA
            target_cls = ModelMQMQA
        return target_cls

    def __getnewargs_ex__(self):
        return ((self._dbe, self.components, self.phase_name,), {'parameters': self._parameters_arg})
[snip]
```
When I test locally it does the correct thing (after I remove the model keywords from the MQMQA tests of `equilibrium`). If the user truly wanted to override the behavior for MQMQA, then in their subclass they can just write
```python
class MyModelSubclass(Model):
    @classmethod
    def dispatch_on(cls, *args, **kwargs):
        return cls
[snip]
```
This would still require them to pass the subclass in the model keyword argument. If they want ModelMQMQA to continue overriding for MQM phases, they would just not define `dispatch_on`, which seems like a sane default to me. It's not as fancy as a metaclass approach, but it's comparatively simple, preserves backwards compatibility, and provides good UX from a least-surprise perspective.

Reply:
My nit with this way of doing it is that now `Model` has to know about all other models to get the magic to work and custom user models will never be able to hook in and get that magic without changing `Model` (and that's true for any dispatcher without registration functionality). All the approaches have tradeoffs.

I didn't want to commit to introducing any magic in this PR, because that sets the precedent for how users should expect to interact with new models. But my experience shows that you have good judgment about how APIs should look, so I'm happy to defer to whatever option you prefer 😄

Reply:
I'm also open to renaming to something like `_dispatch_on` to denote a private API, if that makes this approach easier to accept for now.

Reply:
My proposed changes have been pushed to the branch in [caa1520](https://github.com/pycalphad/pycalphad/pull/389/commits/caa1520dbd5d11f2efbfb4ad59c599c90a9cf87d).